### ISSUE TITLE
Go server and JSON networking

### ISSUE THREAD

**Issue by dangranos** [griefly/griefly#375]
Honestly, I consider using a completely separate program for "server" (though it actually seems to be a relay of some sorts and an auth system) at least slightly weird. It's an additional overhead and requires more dependencies, also maintenance.
Next to that server we have the prtocol, consisting of JSON messages. Plaintext JSON messages, which give us even more overhead because you still need to convert data to and from JSON. And it also requires more from server's internet bandwidth.
Both of those needs to be dealt with by integrating the relay part of the networking into the engine itself and JSON protocol replaced by similarly powerful own binary data protocol.

**Comment by kremius** [comment]
That's an interesting opinion.

About separate server:
1. A separate server does not have an additional overhead. Why should there be an overhead?
2. It is easier to maintain two different well-separated projects than one big bloated project.
3. An abstract separate server does not introduce any new dependecies. In our case it does, but that cannot be reason to say "separate servers are bad". It may be kind of reasonable to say "new dependecies are bad", but see below for clarification.
4. Every added dependency should worth itself. In our case the main dependency that was introduced by the server is golang. Now it works just fine: fast, reliable and extendable, so it is totally worth. Maybe something else can be said by @mechmind, because he is the server code maintainer.
5. It is the common practice to implement the server as a separate binary. Often there are different programming teams for client and server. Only sometimes games have a built-in server, but usually such server is done for players convenience and the server is much more primitive than a dedicated server - which exists too as a separate binary.
6. A built-in server may have issues with hosting on a dedicated server machine.

About JSON protocol
1. "Premature optimization is the root of all evil" -- Donald Knuth.
Everything works just fine now. The imminent v0.4 playtest may show some issues, but now there absolutely no reason to think that current protocol may be a bottleneck. Tests shows that it can handle pretty big amount of messages (it should be fine at least for 100 simultaneous clients). Precise data still not harvested though - testing is needed:
![grafana](https://cloud.githubusercontent.com/assets/3407912/21798570/a3a4dc6a-d72e-11e6-8496-7f772ed5dfe7.png)
2. "powerful own binary data protocol". Powerful, own. Pick one. Developing a good powerful binary protocol is a **HUGE** task and a lot of proficient engineers works on that task. It will be much better to use external well-tested library for such task.

**Comment by PJB3005** [comment]
The server makes sense when you're working with the lock step architecture of griefly.

And that's like saying committing suicide is a good choice when you're going to be tortured to death. 

https://www.reddit.com/r/SS13/comments/5eppon/grieving_for_griefly/

The entire networking stack is fucked and basically past the point of no return.

Having a separate binary makes sense, but in basically all those cases the server does something more than act as a funnel to a "master client" (server) that doesn't realize it's anything but a client. Do you honestly think other games have the code for the server binary be entirely different?

Do I need to begin how using a different programming language ON TOP of that is a terrible idea for maintainability? 

Kremius is right on premature optimization though.

**Comment by kremius** [comment]
Hi @PJB3005
Hope your own remake is going well.

Firstly, I encourage you to use more formal and logical language. When someone is using a lot of emotional prases (e.g, "fucked", "tortured to death") it's harder to grasp the meaning behind such words. Incidentally, it may seem that there is no sense at all - only well-folded words without any real meaning.

>The entire networking stack is fucked and basically past the point of no return.

Maybe this will be helpful for you - #261. If you does not understand something then you can ask me - I will try to explain better. Overall your allegations are really strong and not confirmed by arguments. It would be nice if you explain why it is "fucked".

>Having a separate binary makes sense, but in basically all those cases the server does something more than act as a funnel to a "master client" (server) that doesn't realize it's anything but a client.
1. The griefly server does more than just retranslating messages, but indeed it is the main task.
2. Master client is not "server", because it does not handle incoming connections. His specific role is expressed only in couple things (map sending and hash calculation), it can be changed during the game, etc. The master client does not need to know that it is special, so there is no reason to do that.
3. I do not see the reason here why it is bad to make such tool as a separate binary.

>Do you honestly think other games have the code for the server binary be entirely different?

Well... Yes, I do. At least some games for sure have the totally different code for server and client. There a lot of games, and they are made in various ways, so I don't understand why you asking the question.

>Do I need to begin how using a different programming language ON TOP of that is a terrible idea for maintainability?

Yes, you need. Please, explain to me.

Sorry, but you said a lot of mean things about the network protocol, and does not confirm your words by anything.
Your post is:
1. Statement about lockstep.
2. Sentence with your opinion.
3. Link to a really long emotional text without proper structure and with your feelings about the project. It is impossible to extract exactly your arguments from the text.
4. Sentence with your opinion.
5. Somewhat true statement about servers.
6. Emotional question.
7. Emotional question.
8. Sentence with your opinion.

Overall we have 1 & 5. "Griefly is using lockstep" & "Server and client sometimes are merged into one binary, and it wasn't made that way in Griefly".
Please, consider improving your logical constructions, because I cannot comprehend them now.

**Comment by PJB3005** [comment]
All my points are covered in my Reddit post and the comments I posted in response to others down below.

* Lock step doesn't work in SS13. Period.
* Your entire networking stack of "master client has no idea it's important" is a terrible idea that's just sounds like maintainability *hell*.
* Different programming language is maintainability hell when I see no actual reason why you couldn't even have just written the server in C++ too.

**Comment by kremius** [comment]
>All my points are covered in my Reddit post

"Link to a really long emotional text without proper structure and with your feelings about the project. It is impossible to extract exactly your arguments from the text."

>Lock step doesn't work in SS13. Period.

Why?

>Your entire networking stack of "master client has no idea it's important" is a terrible idea that's just sounds like maintainability hell.

Why?

>Different programming language is maintainability hell when I see no actual reason why you couldn't even have just written the server in C++ too.

It was easier to write server on a different language. Each language has it's purpose, and Go works well for servers. Why different programming language is maintainability hell?

**Comment by shelltitan** [comment]
Well considering your last argument Lua bidding would just bring more hell to griefly, because that would be one more language to maintain. Still you advise to support it in your post so I don't get why are you complaining about Go.

**Comment by PJB3005** [comment]
@shelltitan necessary evil.

@kremius 

> So, not looking good. What if I told you their entire, and I mean entire networking system is a mess. Even BYOND is better than this shit.

> The way networking works in Griefly is lock step. You might know this from Factorio. Factorio uses lock step and they manage huge servers! Surely this is a good idea. The way lockstep works is that every client has its own simulation of the game world, and the clients only exchange keyboard events and such. This works until you have to hide something from the client, like who is antag. In Factorio, if you get a desync, your client has to essentially reset and reconnect to the server entirely (these are bugs and their developers are always looking into fixing these!). In Griefly you would be intentionally desyncing at which point to manage such an architecture you’re not lockstep anymore but some in between and to code for it becomes hard and probably a maintainability mess. Imagine the following: A certain theoretical antag type deflects laser guns. Being an antag, no client is actually supposed to know this, or you could get hacks that simply show who is antag. If the game’s lock step, how does everybody’s simulation except the server’s know that the bullet should be deflected if shot at? When asked about this /u/shelltitann said you’d be able to metagame. No actual solution other than “we can change the netcode to thin client like BYOND really quickly”. I will hand it to them however, because they separated representation and core it shouldn't be the end of the world to pull off.

> But a thin client isn’t the solution either because anything like input and mouse events being done server side is retarded and Dantom should be shot for ever thinking it’s a good idea. See my example of click drag construction and why it doesn't work. 100 ms ping means all mouse events are delayed by 100 ms before you see a change. That's awful. So if you do allow client side code execution with specific property replication and RPCs (calling a function directly on the other end of the connection), it’s possible. Issue is Griefly is in no way tailored to this kind of system, and to do it comfortably in the same sort of style UE4 does it, they’d need to do humongous amounts of work related to reflection.

> Doesn’t help how Griefly’s concept of how a server works is stupid either. See, you don’t have a regular server. The “server” is a Go based program (yes, it’s in a different programming language, sounds fun to maintain doesn’t it!) that basically distributes the network messages around. Then surely there’s a thing that runs game code right? Yes! It’s a client that’s just special and chosen as master. Said client isn’t designed to act as a server. It’s designed to act like any other client. This is probably due to the lock step system, but even then separating the code base this badly just makes 0 sense! Not to mention that the lock step system should never be used in practice for reasons I’ve outlined above. The result is that you would have to make all clients that aren’t the master client at least a thin client. But your master client still isn’t designed to be a server, it’s a client. What handles connections, bans, etc…? The Go based server. Now your server software is fragmented between 2 processes and programming languages. One is the byproduct of a retarded legacy product that shouldn’t have ever been committed, the other doesn’t even think it’s a server. Also, if they were to truly do an authoritative server/client architecture (like they should!), they’re fucked.

And in response to a comment that tried to point some issues out: 


>>    5 and 6 - thin clients are not doomed to be bad. Interface actions can be rendered and handled locally and only high-level actions will be sent over network. Also, predition and lag-hiding stuff is not even tried because current responsiveness was good enough.

> That's called the authoritative server/client model I suggested and it's gonna require scripting at the least to implement.

>>    7 Go was good choice for server implementation. If you do not like server that way you can always provide an alternative, but it have to be better in terms of stability, speed, user friendliness and maintability in long run to consider replace.

> It might've been with the lock step system, but it really isn't when you're working with authoritative server/client because then it does nothing except turn into a funnel.

>>    Also, lockstepped simulations with one master (~admin) client greatly simplified overall network architecture. The server still can be used as central authority because network protocol supports authorization (for now - rudimentary) and network protocol allows embedding control messages into game event stream.

> Of course, your point? You're ignoring the fatal issue in the OP. It works for games like Factorio, but Factorio isn't SS13.

Anyways.



>>    Lock step doesn't work in SS13. Period.

> Why?

With lock step every client must know everything *and* process everything.

If one of your clients is running on a slow as hell computer from '05 they're screwed because they can't keep up and Factorio's solution to this is to flat out kick them I believe.
Not to mention now you need to tell every client who is antag, etc...
You cannot keep anything secret unless you make certain actions require the client to ask the master client whether the action can be done and I highly doubt your game is designed this way, and even if it would be I would be impressed if you can make a good looking API for it.

>>Your entire networking stack of "master client has no idea it's important" is a terrible idea that's just sounds like maintainability hell.

>Why?

Who handles bans. Who connects to the database. Who handles admin preferences???

Unless you want to make basically all admin actions have special code on the Go server to double check whether the action is valid *and* then modify it too (which is a maintainability hell because now code like admin buttons is spread out hugely), nobody is the boss so who does the boss things?

Who chooses which jobs everybody gets? If you want to have the control go to the master client for special occasions like this you're again introducing a mess in every code that needs to do it at least somewhat.

> Why different programming language is maintainability hell?

I sure do love needing to know multiple programming languages because I wanted to give admins a button to make everybody fart.

**Comment by kremius** [comment]
So, your points in first paragraph (+ first line)
1. Network system in the Griefly is full shit
2. But there are actually examples of games where it actually works (it kind of already contradicts with the first point because my definition of "full shit" is kind of different).
3. There are issues with current network system.

The third point is totally correct - every network system has its pros and cons, weak and strong sides, so Griefly's one isn't an exception. I have already written about issues here: https://github.com/griefly/griefly/issues/261#issuecomment-226990760
You here mention two issues:
1. "If desync is happened then it will be needed to reload the map."
Yeah, if there is a bug then something bad happens. That's like software works. High standarts of coding and unit-tests should reduce the amount of such incidents, but there are no silver bullet. Developers makes bugs, users find bugs, developers fix bugs.
2. "Every client knows everything so cheating is possible"
Here some more questions should be asked and answered.
**Is it a serious issue?**
Well, it allows to create maphacks, and discover various hidden stuff (like who is traitor), so probably yes.
**Does it make the game unplayable?**
No, it does not. You can play the game just fine. Just sometimes there might be suspiciously insightful players which acts like they know something extra. SS13 is a roleplaying game, so cheaters should not break immersion or they will be punished. That seriously limits the number of possibile actions for cheaters. Sure, a smart enough cheater still will find a way, but that's not so easy.
**What is with other games that uses such system?**
It is a quite popular system. Most RTS games use/used it (some of them were really competitive and the cheating issue was a lot more serious for them - like Stracraft, Warcraft 3). Factorio uses it, and actually Factorio is much more similar with SS13 then most games. Very complicated world, many things can process by themselves, etc. So even if we assume that it is not the best approach it cannot be denied that it will work.
**Can this approach be changed in future?**
Yeah, it is possible. Representation & game parts should be separated and then we will come to BYOND-like system with a thin client.
**Maybe it just should be done now?**
It is possible, but not easy. It surely will request at least one full milestone for implementation, a lot of hard work and etc. It introduces a lot of other nuances and possible issues (e.g. the server bandwidth usage will be much bigger). Now Griefly does not have cheaters - because Griefly even does not have players now, so no reason to make it right now. Probably it may be planned as one of the core feature of 2.0 version or something like that.

Does not sound like "full shit" yet for me. 

In the second paragraphs you are speaking about the possible solution (thin client) of the cheating issue and really negative about that solution. The reason is "there will be a lag between a player action and a representation of the action". Also you are saying that 100 ms delay was so big and "awful", that you have to give up on some click & drag construction feature.

1. Most real-time games have such lag - that's how the network works. They just hide it by 
animations and such things. When you cast a spell it won't cast immediately. Some initial animation of the spell will be showed, but the real spell starts only after some lag. BYOND have such delay without any lag smoothing, and it works. As far as I know, FPS games tends to have some time reversion on server in order to calculate ballistics, but that approach has a really small field of application. It also gives some opportunity for cheating, but otherwise FPS shooters would be unplayable, so it have to be done.
2. If you are having a big ping in any real time game then you are basically guaranteed to not have a good time. There will be lags, there will be delays. Client may try to compensate for that somehow, but it won't look nice.
3. Even now in Griefly all player actions are delayed by ~100 ms (because even now it is needed to proxy all player actions via server + messages are processed once per 100 ms). It does not feel "awful" for me. Sure, it possible to notice a small delay in various actions (wall crafting, for example), but I don't notice it in the game if I don't pay attention to it. There also wasn't any complaints about "awful" delay from various beta-testers, so it is not only my feel.
4. It is really strange that you gave up on some idea because mouse events are delayed. Usual construction method (clicking on something) should have the same lag with click and drag method. I don't get why one method feels "awful" and other one exists for eternity and is just fine if they have same lags.
5. "So if you do allow client side code execution with specific property replication and RPCs (calling a function directly on the other end of the connection), it’s possible." Griefly already has some built-in kind of lags compensation system in the movement system. SS13 is a plain in a graphical sense game, and it is impossible to use animations for lags hiding. Moreover, it is really complicated game in gameplay sense (a tons of various interactions) - basically, it is a sandbox with a high players interference. The amount of various actions is ridiculous - it causes serious troubles for various lags compensation system methods. So I am not sure it is really possible without losing complexity of the game process.

So still does not sound like "full shit" yet for me.

The third parahraph is about your thoughts about server (and A LOT of your emotions and feelings. Please consider being more professional.). For some reason you assume that "master client" is really server and that it is really bad.
Well, your assumption is wrong. The master client is not a server, and does not work as a server.

Next, some question answers:
>That's called the authoritative server/client model I suggested and it's gonna require scripting at the least to implement.

Scripting have VERY minor connections with the network system. So scripting is not required for anything network related.

>It might've been with the lock step system, but it really isn't when you're working with authoritative server/client because then it does nothing except turn into a funnel.

I don't really understand that you mean by "funnel", but server now does a lot of things. It does nothing except "being server".

>Of course, your point? You're ignoring the fatal issue in the OP. It works for games like Factorio, but Factorio isn't SS13.

Factorio is more similar to SS13 than most games. Big sandbox with a lot of machinery and processing.

So, now your answers for this thread.
>With lock step every client must know everything and process everything.

Yeah, that's true.

>If one of your clients is running on a slow as hell computer from '05 they're screwed because they can't keep up and Factorio's solution to this is to flat out kick them I believe.

Yes, slow machines won't be able to handle the game. It is sad for such people, but it is minor issue - now is 2017, and amount of such machines is small.

>Not to mention now you need to tell every client who is antag, etc...
You cannot keep anything secret unless you make certain actions require the client to ask the master client whether the action can be done and I highly doubt your game is designed this way, and even if it would be I would be impressed if you can make a good looking API for it.

Cheating issue was reviewed in details above.

>Who handles bans. Who connects to the database. Who handles admin preferences???

Go server handles all these stuff!!!

>Unless you want to make basically all admin actions have special code on the Go server to double check whether the action is valid and then modify it too (which is a maintainability hell because now code like admin buttons is spread out hugely), nobody is the boss so who does the boss things?

It is not needed to create new code for every new admin action on server. For admin verbs it will be needed to add special message type (once!) on server with some param. When the server receives such message, it broadcasts the message only if it was received for admin account. That's way an every new admin action (e.g. explosions, item creation) can be coded only in clients. Some actions (bans, for example) will require additional coding in the server, but amount of such admin actions is pretty limited.

Anyway, admin actions is like the most useless part of the game, so I don't know why you point at it like it is something **REALLY** important.

>Who chooses which jobs everybody gets? If you want to have the control go to the master client for special occasions like this you're again introducing a mess in every code that needs to do it at least somewhat.  

Each client calculates for itself which which mob getting which job. They do it simultaneously, and it does not differ from any other game process and action. It is not needed to know anything from the master client in order to choose the occupations.
Master client should be asked only in two cases: when the authoritative map hash is needed and when the authoritative map is needed.

>I sure do love needing to know multiple programming languages because I wanted to give admins a button to make everybody fart.

1. As I already said, it won't be needed to change anything on server most the of time (including "everybody farts" button).
2. In order to change anything in some source code the most important thing it is knowledge of the code structure. The language does not really matter here (at least while languages are somewhat similar). Go is simple language, and in order to make any change in the code won't be needed to learn it for many years. It will be needed to understand the structure of the server code.
3. If you do not love something then you may not do it. Ideally there would be a group of server developers and a group of client developers. For example, now when it is needed to implement a new feature in the server I does not write it myself. I ask @mechmind and we discuss the feature and then he implements the feature.

And also,
>necessary evil.

Scripts aren't necessary. They are just convenient for a certain group of people.


You have said a lot of untrue or ambiguous things and generously seasoned your post with a lot of mean and rude words and phrases. Someone could even say that you are "stupid" or "blinded with own incompetence", but I will assume that you just really concerned about future of the SS13 and that's why you are so emotional. Please, chill out and don't use your emotions in a technical discussion.

**Comment by PJB3005** [comment]
> But there are actually examples of games where it actually works (it kind of already contradicts with the first point because my definition of "full shit" is kind of different).

I'd love to know which supposed game is an SS13 remake that has lock step networking and isn't Griefly. Because it works in some games doesn't mean it applies to SS13.

> "If desync is happened then it will be needed to reload the map."
>Yeah, if there is a bug then something bad happens. That's like software works. High standarts of coding and unit-tests should reduce the amount of such incidents, but there are no silver bullet. Developers makes bugs, users find bugs, developers fix bugs.

That was an example of how badly things break if it happens, not an argument against griefly.

> Also you are saying that 100 ms delay was so big and "awful", that you have to give up on some click & drag construction feature.

You have no idea how awful delay on mouse movement is at all. No instant feedback is already hideous and if you think it's fine because it works in BYOND SS13 it's because you've grown accustomed to the entire thing being a poorly responsive mess, due to the thin client.

And yes, we still have the feature but that doesn't make it feel less awful to use.

> The third parahraph is about your thoughts about server (and A LOT of your emotions and feelings. Please consider being more professional.). For some reason you assume that "master client" is really server and that it is really bad.
Well, your assumption is wrong. The master client is not a server, and does not work as a server.

This is literally arguing semantics. Because I can put a steering wheel on something to rotate it doesn't make it not a steering wheel. The master client and and Go "server" cannot be separated so why bother sort of doing so?

> (and A LOT of your emotions and feelings. Please consider being more professional.)

Of course I'm gonna use emotions when I'm of the opinion you're literally going to ruin the SS13 ecosystem unless you stop right now to refactor the entirety of Griefly, which I know you won't do.

> Scripting have VERY minor connections with the network system. So scripting is not required for anything network related.

I will be genuinly impressed if you recreate UE4's UHT so you can actually have a good looking API for defining object variables as replicating and such in C++. Scripting languages like Python can implement this kind of stuff with decorators and descriptors without looking awfully.

> Factorio is more similar to SS13 than most games. Big sandbox with a lot of machinery and processing.

SS13 is not Factorio. Period.

> Yes, slow machines won't be able to handle the game. It is sad for such people, but it is minor issue - now is 2017, and amount of such machines is small.

You're still putting extreme amounts of useless load on clients. "Get a better computer" should not be an excuse for you being too lazy to come up with a good networking architecture.

> Go server handles all these stuff!!!

This only enforces my argument of the codebase being fractured between two programming languages and becoming a maintainability hell.

> It is not needed to create new code for every new admin action on server. For admin verbs it will be needed to add special message type (once!) on server with some param. When the server receives such message, it broadcasts the message only if it was received for admin account. That's way an every new admin action (e.g. explosions, item creation) can be coded only in clients. Some actions (bans, for example) will require additional coding in the server, but amount of such admin actions is pretty limited.

You have no idea how deep down the rabbit hole this gets. What about stealthmins, what about admin actions that should be silent and the client shouldn't be able to directly know about directly? Same issue as cheating really.

> Anyway, admin actions is like the most useless part of the game, so I don't know why you point at it like it is something REALLY important.

"We'll cross that bridge when it comes to it" is not always a good approach to designing things. Stop pretending like it is.

> Each client calculates for itself which which mob getting which job. They do it simultaneously, and it does not differ from any other game process and action. It is not needed to know anything from the master client in order to choose the occupations.

How are you gonna do deterministic random chances with this architecture? If it's anything more a single function to call without needing extra arguments for networking you blew the API design.

> 1. As I already said, it won't be needed to change anything on server most the of time (including "everybody farts" button).
> 2. In order to change anything in some source code the most important thing it is knowledge of the code structure. The language does not really matter here (at least while languages are somewhat similar). Go is simple language, and in order to make any change in the code won't be needed to learn it for many years. It will be needed to understand the structure of the server code.
> 3. If you do not love something then you may not do it. Ideally there would be a group of server developers and a group of client developers. For example, now when it is needed to implement a new feature in the server I does not write it myself. I ask @mechmind and we discuss the feature and then he implements the feature.

This seems more like saying the fractured nature is more workable than I'm making it out to be, not that it's a good idea in the first place.

> Scripts aren't necessary. They are just convenient for a certain group of people.

I'm not sure how I'm gonna respond to this.

You're either absolutely retarded and fail to see why SS13 is even alive today, or you're just trolling me at this point.

If you honestly truly think that scripting isn't necessary then sorry but you genuinly need to ***wake the fuck up***.

It only proves that Griefly also suffers from the exact same fucking issue as every other remake, and that is that nobody that starts a remake even seems to understand how SS13 it works, and I'm not talking code wise here either.

You have no understanding of SS13's ecosystem at all. You're seriously not fit to design a remake. I would go on *if I didn't do that on that Reddit post already.*

P.S. Did you even read the Reddit post?

**Comment by kremius** [comment]
>I'd love to know which supposed game is an SS13 remake that has lock step networking and isn't Griefly.

And I would like to know which supposed game is an SS13 remake. Something that is still actively developed and isn't Griefly. It is not much sense to compare Griefly with non-existent games.

>Because it works in some games doesn't mean it applies to SS13.

And it also doesn't mean it doesn't apply to SS13. Lock-step is mostly used in complex games with a lot of game objects and a complex gameplay. Just like SS13.

>You have no idea how awful delay on mouse movement is at all. No instant feedback is already hideous and if you think it's fine because it works in BYOND SS13 it's because you've grown accustomed to the entire thing being a poorly responsive mess, due to the thin client.

I think I have an idea. Because I have played a SS13, and I have played other games with instant feedback. If it is possible to get used to it - great! That means that the issue isn't really a big issue :) Maybe you exaggerate or you are too sensitive? Anyway, seems like it is the feature of perception. From my statistics, only a small amount of people thinks that it is a critical issue (basically, only you and you think that everything is critical). It is sad for those people, but it is impossible to make everyone happy.

>The master client and and Go "server" cannot be separated so why bother sort of doing so?

They **are** separated. From the server perspective the master client is nothing more than "client that can be trusted". Such client can be changed, for example. Or we can have more than one such client (well, we can't right now, but overall design allows it)! It may be helpful in some cases (e.g. extreme amount of clients).

>Of course I'm gonna use emotions when I'm of the opinion you're literally going to ruin the SS13 ecosystem unless you stop right now to refactor the entirety of Griefly, which I know you won't do.

It is a really interesting argument.
I don't see how my actions can possibly do anything bad to SS13. It is not like people will be forced to play Griefly - they will play Griefly only if Griefly offers them more in total than BYOND. If Griefly will be really that bad then it will just slowly die and everyone will know for sure that your assumptions were totally correct and that a faithful SS13 remake should be made in a different way. There are a lof of new projects in all areas in this world and 99% just fail for various reasons. It does not "ruin" anything, but it gives valuable information about the area.
Anyway, emotions never help in difficult situations (well, maybe except childhood). A little bit of thinking always helps more.

>I will be genuinly impressed if you recreate UE4's UHT so you can actually have a good looking API for defining object variables as replicating and such in C++. Scripting languages like Python can implement this kind of stuff with decorators and descriptors without looking awfully.

As far as I know UE engine mostly used for various action/fps games, so I am not sure that it will work in SS13 realities without reducing complexity of the game.
I don't know all the inner details of Unreal Engine, so I can't say much about that. I will be happy to see SS13 remake on that engine!

>Factorio is more similar to SS13 than most games. Big sandbox with a lot of machinery and processing.
>SS13 is not Factorio. Period.

So you I give you the statement and arguments and you give me "your statement is not true" without arguments? I will count that as my win :)

>You're still putting extreme amounts of useless load on clients.

It is not useless because all clients use the produced information. Yes, all clients do the same work, but they do not have to pay the delivery price in exchange.

>"Get a better computer" should not be an excuse for you being too lazy to come up with a good networking architecture.

1. If something does not work for 1% of people but gives you enough profit then it should be used without a doubt. You may have working xor perfect result.  
2. I wish that there was a good networking architecturue! I have thought a lot and I think that there are only two possible network architecture for SS13 game - lockstep and BYOND-like client. Various approaches that are used in other games work only if SS13 will be seriously cropped in terms of features.

>Anyway, admin actions is like the most useless part of the game, so I don't know why you point at it like it is something REALLY important.
>"We'll cross that bridge when it comes to it" is not always a good approach to designing things. Stop pretending like it is.

I'm saying that bridge is not that needed and not so important. Most admin features give good experience only to admins. 

>How are you gonna do deterministic random chances with this architecture? If it's anything more a single function to call without needing extra arguments for networking you blew the API design.

https://en.wikipedia.org/wiki/Pseudorandom_number_generator
It was done a long time ago.

>This seems more like saying the fractured nature is more workable than I'm making it out to be, not that it's a good idea in the first place.

It was easy to do, it is easy to work with it now, and it seems like there woundn't any major issues in future (but who knows!). Seems like a good idea for me.

>You have no understanding of SS13's ecosystem at all.

I think I have **full** understanding of SS13's "ecosystem". I started to play the game since 2010. I was a player, a coder, an ingame admin, a server admin, a forum shitposter, etc. It is like an ecosystem in some industrial city.
It seems like you think that having tons of coders is really good and important (that's why you are so obsessed with scripting). It is wrong - it is needed to have a dedicated organized group of people in order to make good projects. This group can be pretty small. Even one man sometimes could make a lot more than plenty of people (examples: games - Dwarf Fortress, SS13 builds - Lifeweb).

>You're seriously not fit to design a remake.

I would better be the man who will make a remake, not the one who designed the perfect, but an unrealistic game.

>P.S. Did you even read the Reddit post?

Yes, I did. Your writing is looking nice but in reality it is like
1. Make a false or controversial assumption
2. Do a lot of whining and crying based on that assumption with little or no arguments at all. Sometimes the next assumtion is based on the previous false assumption.
3. Goto 1.

What should I do in response? I do not know for sure. When someone is wrong on so many levels it is hard to have a worthwhile discussion.
I hope my words here will help third-party observers to understand a very strange nature of your accusations.

**Comment by monster860** [comment]
Heres why this is gonna fail.

Look at the list of ss13 servers. Each one has their own wiki and their own codebase. They are all different. You just click on a server and it downloads some graphics abd sound and your game starts. With this, you cant do thar. Unless of course you want it to be "click button, download and run a questionable exe file and you now have a virus for joining this new exciting looking server."

**Comment by Lourens-Rich** [comment]
Don't turn off your UAC. You must to know it.

**Comment by PJB3005** [comment]
@Lourens-Rich UAC doesn't protect something from encrypting all my personal files then making me pay a ransom.

**Comment by shelltitan** [comment]
It's not like anyone of the other remakes attempt offered that it will support multiple server from the beginning and talking about a feature when core things aren't even finished, writing post on your personal beliefs just because you cannot stand that something doesn't follow your personal preferences are so useful that you should just keep doing it. You can simply think that why this going to fail but if you are so sure that this is going to fail then why are you wasting your time to tell it monthly on reddit or here or in discord.

**Comment by Lourens-Rich** [comment]
For this invented the anti-virus software. (If you love open unknown and untrusted applications from user which contain valuable or personal files...) And don't forget that any app can be hacked. (Byond too)

**Comment by PJB3005** [comment]
@Lourens-Rich Anti-virus won't do shit here.

Yes, no software is infallible, and definitely not BYOND, but you have to at least find a remote code execution exploit which is really hard. Requiring binaries per server is basically opening the fucking flood gate to malware.

**Comment by Lourens-Rich** [comment]
@shelltitan, becouse @PJB3005 hysteria...

**Comment by PJB3005** [comment]
@Lourens-Rich nice meme.

**Comment by Lourens-Rich** [comment]
@PJB3005, you can check this source code

**Comment by PJB3005** [comment]
Yes, and I'm sure that average joe is smart enough to check the code of any server he connects to. And even then he'd have to compile himself, because binaries from the server could be a patched version with malware even if the public source is different. And Griefly's license, MIT, doesn't even require open source.

No, you can't check the source code.

**Comment by PJB3005** [comment]
Unless every code edit is audited and signed by a trusted team and *THEY* provide the binaries, it's technologically impossible to prevent malware if you're not putting custom code behind a scripting layer.

Such an auditing system would require immense amounts of man power, trust, etc... which would also make it impossible.

**Comment by kremius** [comment]
@monster860 
So, I can find two points in your message:
1. It will be harder for players to get new/another versions of the game.
2. It will be harder for developers to promote their new builds.

Reasons are the same for both points: manual download & some malicious behaviour possibility.

First of all, I would like to present a class of games which has such flaws in their design, yet they are still quite popular. It is rogue-like games. So these issues are not really critical for a successful game or even a big game community :)

Why so? I think it is because if something results in a greater player experience then minor issues like that do not actually matter. Good features should outweigh such obstacles.

However, I might be wrong and these issues will be much more serious than I expect, and you will be right.
Fortunately, there is always a Plan B - a 2.0 network system (I have mentioned it above several times). It should solve these issues.

So thanks for pointing out these issues. I hope I have clarified my point of view. If there are any questions then feel free to ask me.

**Comment by Lourens-Rich** [comment]
Did you think that after so many remakes, other people to be play on all griefly clones (forks)? No. Peoples usually using only 2-3 popular variations of anything. Mind you, you haven't offered any real advice. I have one. @kremius can create a list of approved future similar projects.

**Comment by ShadowCreator** [comment]
Off topic but @kremius what software are you using for those tests near the beginning of this issue? It looks badass.

**Comment by kremius** [comment]
@ShadowCreator
It is [Grafana](http://grafana.org/).

**Comment by PJB3005** [comment]
@Lourens-Rich there at at least 10 different forks of SS13 with active servers right now.

"List of approved future similar projects" is exactly what I said with auditing and it's not possible.

**Comment by Lourens-Rich** [comment]
@PJB3005, excuse me, but i think it is better to have such a project than nothing. I want to run SS13 native for my linux, but can't.

**Comment by kremius** [comment]
Yes, a list of approved projects isn't a very promising idea. Because 1) hard to maintain 2) actually is not needed. If your project is good then just build up some trust and the playerbase will appear despite of the download issue. That's how things works outside of the BYOND and they work fine (rogue-like games is the classic example, but actually any application on the PC works that way).

And after all, @PJB3005 and others, you are worrying too much about things that are too far away. "The game is actually enjoyable and playable" point of the development is 2-3 years away. If Griefly will achieve this state then I will already be happy. The milestone "make the game easily modifiable" is so far away that it is hard say anything for sure about the milestone. Too many things can happen, too many reasons can be changed. SS13 can die, BYOND can be heavily optimized _(lol)_, Griefly will be popular despite of the harder modding, etc. Who knows that will occur it this erratic world?

It is not like the network cannot be changed later or it will be impossible to add scripts later. So it is better for me to focus on more actual tasks.

**Comment by PJB3005** [comment]
> Who knows what will happen

No, this timeline WILL happen and that's exactly why I need you to wake the fuck up before you ruin SS13 development.

> It is not like the network cannot be changed later or it will be impossible to add scripts later.

Rewriting the entire networking and content stack is not as simple as "oh yeah we can do it later".

**Comment by kremius** [comment]
>this timeline WILL happen

Are you a time-traveler? Because seems like you believe in my project more than I do. I am kind of flattered, but I gave you the reasons why that may not happen.
>you ruin SS13 development.

I am not sure that you mean by "development" and how someone can kill something in SS13 by creating a little pet project in his spare time.
>Rewriting the entire networking and content stack is not as simple as "oh yeah we can do it later".

Rewriting? Maybe. Good thing that I don't mean a rewriting. I mean an enhancement of the existing system. And yes, it will not be needed to move already existing content to scripts if scripting will be added.

**Comment by ModernViolence** [comment]
I strongly do not want to shitpost...
But this thread made my evening. Thank you, everyone!
@kremius: Good luck with your project!

**Comment by Quehadt** [comment]
>It seems like you think that having tons of coders is really good and important (that's why you are so obsessed with scripting). It is wrong - it is needed to have a dedicated organized group of people in order to make good projects. This group can be pretty small. Even one man sometimes could make a lot more than plenty of people (examples: games - Dwarf Fortress, SS13 builds - Lifeweb).
I wish not to start broad arguing conversation but I feel need to make it clear:
While DF Is bright example of hard labour of one man which have accomplished complex game with crystal clear structure, Lifeweb is rather work of multiple people with ripped parts of other games( sounds mainly). Main concept is rather good but paranoid behavior of one of it main creators and money-driven community policy is drastic impact.

**Comment by bauen1** [comment]
When bandwidth becomes a problem, http://msgpack.org/ is just like json but a bit smaller.

**Comment by kremius** [comment]
@bauen1 Looks quite promising. Another option is protobuf (https://github.com/google/protobuf), but at first glance I like msgpack more, because protobuf seems like much more bulky.

**Comment by dangranos** [comment]
And bumping this issue up with a return to first point of this issue, namely, the Go "server".

Quoting kremius' comment on this
> About separate server:
> 1. A separate server does not have an additional overhead. Why should there be an overhead?
> 2. It is easier to maintain two different well-separated projects than one big bloated project.
> 3. An abstract separate server does not introduce any new dependecies. In our case it does, but that cannot be reason to say "separate servers are bad". It may be kind of reasonable to say "new dependecies are bad", but see below for clarification.
> 4. Every added dependency should worth itself. In our case the main dependency that was introduced by the server is golang. Now it works just fine: fast, reliable and extendable, so it is totally worth. Maybe something else can be said by @ mechmind, because he is the server code maintainer.
> 5. It is the common practice to implement the server as a separate binary. Often there are different programming teams for client and server. Only sometimes games have a built-in server, but usually such server is done for players convenience and the server is much more primitive than a dedicated server - which exists too as a separate binary.
> 6. A built-in server may have issues with hosting on a dedicated server machine.

So...
1. Okay, I admit, there is no network overhead, or next to none. But consider this, that in most use cases this "server" is launched alongside the game and uses up resources. Sure, it's negletable on a small server, but the more clients you have, the more things this thing must route between itself and the master host, which is actually the **real** server in here. Point is, it's separate program that host of the server will have to deal with and set up and launch. For huge servers it might be fine, but for a quick game between friends that would be just a pain in the ass. Nevermind compiling it.
2. Yes, it is. At the cast of it being two separate codebases, that only have a common protocl in between and are maintaned by **two separate people**.
3. It's not a server. It has zero gameplay logic on it. It's just a dumb relay and authorization program.
4. And this thing is the only thing in the project that requires it. But for developers/hosters that requires installing a Go compiler and setting it all up and so on. Unnecessary and avoidable by integrating it into C++ engine itself.
5. Yes, and the common practice to have most/whole gameplay logic on such server. You answer here is invalid, because it's impossible to call this "go server" blob anything close to actual game servers. Even Factorio that uses similar lockstep networking structure has what is practically graphics-less client as a server. Which should be the dedicated server in this case. Not a weird program that only coplicates matter of delivering things between clients.
6. See point 5, make a dedicated server that is a graphics-less client. Should not be much of a problem.

**Comment by tuesdayMarch** [comment]
>2). Yes, it is. At the cast of it being two separate codebases, that only have a common protocl in between and are maintaned by two separate people.
>4). And this thing is the only thing in the project that requires it. But for developers/hosters that requires installing a Go compiler and setting it all up and so on. Unnecessary and avoidable by integrating it into C++ engine itself.
>6). See point 5, make a dedicated server that is a graphics-less client. Should not be much of a problem.

2). Doesn't seem to be an issue at the moment: "Premature optimization is the root of all evil." If it's not an issue, why spend time on it? Better to get an MVP/ base finished first before worrying about that.
4). This is a minor issue. Go is easy to setup and if it comes down to it, an installer can be easily created.
6). This would be convenient.

**Comment by dangranos** [comment]
2. Screams"maintainability hell". And, IIRC, this thing would supposedly
also somehow deal with admin command. It would be pain to implement in a Go
server, but pretty easy to do so if it was interested into engine.
4. That's still one more step. The easier the compilation and running is,
the more time developer will spend on working instead of messing with go
and thinking "who made this?!"

**Comment by PJB3005** [comment]
I'm against the go server here.

1. It's called threading. Having it in another process is more overhead because of serialization and the alike. Also Go's slower than C++ and having it in another process doesn't magically remove the CPU usage. It still needs to be done.

2. No, it's not. Change Auth code? Two languages, two codebases, etc...

3. The concept of a separate server is still dumb either way.

4. Why couldn't you write code like this in C++? It's a lot faster than Go.

5. Like dangranos said, it's common to have this as a separate binary, but still just link against a shared DLL or something, *not* to have a separate binary on top of it. Can you show me *any* game that does it like you do?

6. That just shows you have no idea what you're talking about honestly.

**Comment by kremius** [comment]
1. I don't think you understand how Griefly network system works. The only piece of info which is requested specifically from a master client is a snapshot of the world. It is requested only when new client is connected. Also there might be multiple master clients and the server may choose a suitable one.
2. Seems like you are agree with me here, so I don't know what to say. Maybe you are afraid that @mechmind may vanish? Well, in that case I can say that Go is designed specifically as a simple language and I will be able to handle it :-)
3. https://en.wikipedia.org/wiki/Server_(computing)
The current server provides a service - it relays messages. It is a simple task, but it should be done. If it somehow interferes with your concept of servers maybe you should expand them a little bit.
4. If a developer cannot manage to install something so easy as Go package (it is developed by Google so everything is made cool) then probably he won't be able to contribute anyway.
5. (and 6) That's probably one of the most minor issues possible. If the game will be good it will be played even it is needed to setup server for it. If the game will be bad (or, realistically speaking, just unfinished) then it does not matter. Moreover, it is possible to make button in the launcher which will start a local game server.

**Comment by shelltitan** [comment]
@PJB3005 MMOs mostly have a separate binary for server and sometimes they have more than one server related binary.

**Comment by PJB3005** [comment]
@kremius 
1. How or not it works here is literally irrelevant against the performance point. Also I'm sure those snapshots are all that auth logic and stuff too.
2. That... doesn't make it easier to maintain, at all. I don't know what you're on but that's just wrong. C++ is hard enough for most people and then they have to learn Go too?
3. That's just semantics.
4. What about the dummy who can barely use Git and just wants to add his mcnuggets with szechuan sauce to the game? Wanna know why they can code and test it in BYOND? *Because the compiler and editor are already installed and easy to use*.
5 & 6. "it can be worked around to become irrelevant" is not an argument, especially because this'll essentially remove any point to having it separate, so why even have it separate at that point?

@shelltitan 
MMOs are gigantic complex beasts and can't even be used as arguments here because they're literally on another level.

**Comment by dangranos** [comment]
Could you please stop calling this a server and call it a relay, because
that's all that is.

You are providing Wikipedia definition of a server, but keep in mind Thai
in game development server is what is going all the game logic. And right
now, that's peers/clients. And the go relay is just a unnecessary
overcomplication of networking/sync protocol for griefly.

As for the MMO statement... This is just plain stupid. MMOs are using
sharding, technique when a single world is actual composed of several
gameplay servers, or several dozens of them, making up a single
network/world usually serving a region, and all those servers are connected
together by a gateway. Yes, gateway relays messages, but it also tracks
client and makes it transition between servers seamlessly, or more or less
without much of a fuss on client. And they serve least dozens clients, but
most likely severally hundreds, of not thousands of clients. Your argument
is invalid because those are not comparable.

**Comment by tuesdayMarch** [comment]
@PJB3005 
2). Go is a very easy language compared to C++. If a programmer wishes to work with the engine, he may decide to learn how to use all of its tools. If not, he can stick with Byond and DM. It's up to him alone.
4). Then he may stay with Byond if he doesn't want to learn how to use Griefly. No one is stopping him, only his ambition. 
5 & 6). You forget this is not a forum for argumentation. The design decisions are up to Kramer alone, he is the maintainer of Griefly, no one else. You can make your points and be persuasive, but he has the right to not consider them and do as he wishes.

If you see a problem with the current system, make your own and make a pull request.

There's no point in being visceral here.

**Comment by shelltitan** [comment]
You can always fork and follow your ideas.

**Comment by PJB3005** [comment]
@tuesdayMarch 
2. The point of a remake is to replace SS13. You can't just say "he can stick with BYOND" because it'll come down to switching *eventually* (if griefly succeeds), and when it does there'll be no solution because you didn't think it through and are now stuck with having to refactor the entire code base.
4. See point 2.
5. 
> You forget this is not a forum for argumentation. 

When was this said? Judging by the past conversation it very clearly *was*. If Kramer (that's their name?) wants me to stop he should give the word.

> The design decisions are up to Kramer alone, he is the maintainer of Griefly, no one else. You can make your points and be persuasive, but he has the right to not consider them and do as he wishes.

I'm literally just trying to talk sense into him, because the current system *does not work*. Do you think I'm trying to be destructive or anything here? If Griefly is gonna stick around the least I can do is try to prevent it from going downhill even further.

> If you see a problem with the current system, make your own and make a pull request.

This directly contradicts your previous statement, because if it's up to Kramer whether or not to merge it, and his opinion is pretty damn clearly fixed, *there's a 0% chance of it getting merged*. Thus I'm arguing to try to talk sense into him. (note: I'm not coding it either way)

@shelltitan
> You can always fork and follow your ideas.

The last thing a remake needs is a fork to split development. This just comes off as passive aggressive and throwing the cliche "you can fork!" whenever there's criticism about an open source project.

**Comment by shelltitan** [comment]
@PJB3005 When the SS14's old network coder had to explain to you that there isn't better way to synchronize than delta snapshots with fixed frame server to not delete the his code, then I think you cannot be considered an expert in networking area. I don't want to pick on you but what problem do you have with Go when you use C#, performance isn't your preference when it's about choosing languages.
 
https://benchmarksgame.alioth.debian.org/u64q/compare.php?lang=csharpcore&lang2=go a benchmark to compare C# and Go.

**Comment by PJB3005** [comment]
First of all, all volundr said was "don't tear the old one out, fix it.
It's trash but it works" he even admitted that a more sane system like we
were discussing was better, but would take too long to implement.

Also, I'm arguing at Go vs C++ because that's what it's about here with
Griefly. C# for SS14 is a great choice either way for reasons I won't get
into here, but if you want an explanation hop onto SS14's Discord.

**Comment by mechmind** [comment]
I followed this discussion and as far as i can tell there are several issues you have with server:
- It's hard to just start a server when playing with friends;
- Server implemented in go;
- We call it "server".
Lets see what we got here.

The first and real issue can be solved by a small amount of additional logic in launcher so it can start server automatically and comparable amount of additional work in server and client so server can work in LAN and through NAT. After implementing this logic game can be hosted from launcher seamlessly. These changes do not require complete replacement of current codebase.

The second issue is language of choice for server. Argument states that having server implemented in another language leads to maintability mess and requires learning another language when adding content to Griefly. This is not true. Wire protocol does not require fundamental changes when adding content. For example, shelltitan was able to implement a lot of game features last year without touching server code at all.
Also rewriting current server in C++ will not be as easy and straightforward as one can assume. Most likely there will be a need in asynchronous network server and state machine so server can support hundreds of clients without issues. And debugging asynchronous network servers in C++ is a tedious task.
Besides that, current server does not run simulation on server side so there are no need in powerful hardware to run it.

And to the naming issue. SS13 runs simulation on server and only representation on client but this is just an implementation detail. If griefly network architecture splits into "relay", "simulation" and "representation" nodes there likely will be a rename but current naming is clear enough.

**Comment by kremius** [comment]
@PJB3005 

>I'm against the go server here.

Well, seems like you against everything in Griefly, so why bother to even mention this?

> 1. It's called threading. Having it in another process is more overhead because of serialization and the alike. Also Go's slower than C++ and having it in another process doesn't magically remove the CPU usage. It still needs to be done.

I don't really get you, sorry. Do you think that making server in C++ magically removes threads, processes or serializations? Why do you think threading is bad? Well, I am totally not sure.

> 2. No, it's not. Change Auth code? Two languages, two codebases, etc...

Well, such things hard to explain to be honest. It is from my personal experience - if it is done right then two separate codebases (even if there will be a little bit extra work) are easier to maintain.

> 3. The concept of a separate server is still dumb either way.

Well, you think it is dumb, I think it is not.

> 4. Why couldn't you write code like this in C++? It's a lot faster than Go.

Because it was simpler to write the server in Go. If Go will be too slow then the server will be rewritten in C++, but only then. I am not sure it will be possible to really reach limits of the Go language in the server on the SS13 level.

> 6. That just shows you have no idea what you're talking about honestly. 

Hmmm... Hmmm :-)

>How or not it works here is literally irrelevant against the performance point. Also I'm sure those snapshots are all that auth logic and stuff too.

From performance point of view there aren't any issues. And no, it is not about auth logic and stuff too. It is just world snapshot.

>That... doesn't make it easier to maintain, at all. I don't know what you're on but that's just wrong. C++ is hard enough for most people and then they have to learn Go too?

Go is much easier than C++, so if they will be able to C++ then I am sure they will handle Go. Moreover, the relay-server is quite simple by it's nature, so there should not be much code to support.

>That's just semantics.

Well, semantics is basically the most important part of a word.

>What about the dummy who can barely use Git and just wants to add his mcnuggets with szechuan sauce to the game? Wanna know why they can code and test it in BYOND? Because the compiler and editor are already installed and easy to use.

Frankly speaking, such people are not really useful. 100 weak coders cannot outdo even one average coder. If there will be small dedicated group of people that will be enough for creating a good game.

>5 & 6. "it can be worked around to become irrelevant" is not an argument, especially because this'll essentially remove any point to having it separate, so why even have it separate at that point?

1. It is easier to implement a small relay server as a separate project.
2. It is easier to maintain such server.

If something can be made a separate it basically should be made that way. Now the purpose of the relay server is clear - it is just a relay server which broadcasts messages. It depends only from protocol structure - and the protocol (kind of) is well-defined. So it works just great as a separate executable - and it is common sense to do it that way.

@dangranos
>Could you please stop calling this a server and call it a relay, because that's all that is.

Well, I will call it "relay-server" as a compromise. Because it is definitely a relay, and definitely is a server. The wiki definition is area-agnostic, so it can be used in game development as well. The server is the stuff that receive connections.

>And the go relay is just a unnecessary overcomplication of networking/sync protocol for griefly.

You don't understand protocol then!

And @PJB3005 again!
>2. The point of a remake is to replace SS13. You can't just say "he can stick with BYOND" because it'll come down to switching eventually (if griefly succeeds), and when it does there'll be no solution because you didn't think it through and are now stuck with having to refactor the entire code base.

If Griefly will succeed then there are no issues at all. Because if someone won't be able to move from BYOND to Griefly, but the game will still work then that coder was not needed really. If Griefly will not be able to grow because of lack of coders then it will not succeed. 

>Also, I'm arguing at Go vs C++ because that's what it's about here with Griefly. C# for SS14 is a great choice either way for reasons I won't get into here, but if you want an explanation hop onto SS14's Discord.

Then you should not use the performance argument because it does not matter.

>When was this said? Judging by the past conversation it very clearly was. 

So, if it should done explicitly:
@PJB3005, **it is not a forum for argumentation!**
I am really appreciate questions which people sometimes ask (because it helps to understand the future of the project better) and try to provide proper answers to them - but sometimes people just cannot understand. I spend too much time trying to answer you properly and you just don't understand basic things! It seems like you have never worked on a real project or anything except BYOND seriously. I maybe wrong, and I am really sorry then, but that how things look for me.
Anyway, seems like I have never understand you and you will never understand me, so you can stop try to persuade me with the same arguments over and over. I can recommend you to **focus on your own remake**.

**Comment by PJB3005** [comment]
> Well, seems like you against everything in Griefly, so why bother to even mention this?

Now that's just passive aggressive.

> I don't really get you, sorry. Do you think that making server in C++ magically removes threads, processes or serializations? Why do you think threading is bad? Well, I am totally not sure.

I meant that you should just use a second thread on the master "client" in place of the Go server.

> From performance point of view there aren't any issues.

I didn't bring it up. Sure there's no issues but saying there's no overhead at all is a lie.

> the relay-server is quite simple by it's nature

It's gonna quickly become larger as more sophisticated admin code begins to exist and it will hurt.

> Frankly speaking, such people are not really useful. 100 weak coders cannot outdo even one average coder. If there will be small dedicated group of people that will be enough for creating a good game.

Until they learn programming proper and become an important coder after adding their mcnuggets. You're alienating a large group of people.

> If something can be made a separate it basically should be made that way.

Just because something should be separate doesn't mean it must be in an entire different executable or project. Unless your code is absolute spaghetti it's easy to modularize things and keep things sane in a single project.

> You don't understand protocol then!

Don't pretend like this server is a good thing in general. It's just a terrible crutch on top of the lock step networking.

> If Griefly will succeed then there are no issues at all. Because if someone won't be able to move from BYOND to Griefly, but the game will still work then that coder was not needed really. If Griefly will not be able to grow because of lack of coders then it will not succeed.

Again, alienating a ton of potential coders.

> Then you should not use the performance argument because it does not matter.

I didn't bring it up.

> it is not a forum for argumentation!

Alright, fine then.

> you just don't understand basic things!

I spend too much time trying to answer you properly and you just don't understand basic things!

> It seems like you have never worked on a real project or anything except BYOND seriously.

I actually haven't but you don't have to have programming experience to see that lock step is awful. *cough* factorio multiplayer is laggy as hell *cough*

> Anyway, seems like I have never understand you and you will never understand me, so you can stop try to persuade me with the same arguments over and over.

Unsurprisingly I can say the same.

> I can recommend you to focus on your own remake.

Alright then. Good luck to you too.
